File: ruby/thousands_separator.rb
class ThousandsSeparator
  # Method to check the value conditions
  def check_number(number)
    # only number negative/positive is accepts or !== Sting/Integer
    return '0' if (!number.is_a?(String) && !number.is_a?(Integer)) || number.to_s.scan(/^(|-?\d+)$/).empty?

    # number length < 3 does not need treatment
    return number.to_s if number.to_i.abs.to_s.length <= 3

    # return raise if number > 15
    return raise ArgumentError, 'Number length is > 15' if number.to_i.abs.to_s.length > 15

    # return true
    true
  end
end

# Solution 1
class ThousandsSeparatorRegExp < ThousandsSeparator
  def format(number, separator = ',')
    value = check_number(number)
    if value != true
      return value
    else
      format_number(number, separator)
    end
  end

  private def format_number(number, separator)
    # Solution usigin reverse + scan(regex) + join
    # number.reverse.scan(/\d{1,3}/).join('#{separator}').reverse
    # Does not seem is efficient?!

    # Alternative with gsub
    number.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, "\\&#{separator}")
  end
end

# Solution 2
class ThousandsSeparatorArray < ThousandsSeparator
  def format(number, separator = ',')
    value = check_number(number)
    if value != true
      return value
    else
      format_number(number, separator)
    end
  end

  private def format_number(number, separator)
    # Alternative with array structure
    number = number.to_s.chars.reverse.each_slice(3).map { |num|
      num.join
    }

    # Remove first comma in array if number contains negative symbol
    # e.g. ["654", "321", "654", "321", "-"]
    if number.include?('-')
      number.delete('-')
      number[-1] = number.last + '-' # add comma on string (last)
    end

    # return string 
    number.join(separator).reverse
  end
end

# Solution 3 ~> Idea by: Cary Swoveland via https://stackoverflow.com/a/26682481
class ThousandsSeparatorLoop < ThousandsSeparator
  def format(number, separator = ',')
    value = check_number(number)
    if value != true
      return value
    else
      format_number(number, separator)
    end
  end

  private def format_number(number, separator)
    value = number.to_s     # string
    value_size = value.size # size

    # Loop with for
    # i = 1; i < value_size; i += 3
    # for num in (3...value_size).step(3)
    #   value.insert(value_size - num, separator)
    # end

    # Alternative loop with step
    (3...value_size).step(3) { |num|
      value.insert(value_size - num, separator)
    }

    # Resolve problem with negatives
    if value.include?('-,')
      value = '-' + value[2..-1]
    end

    # return new string
    value
  end
end

# Solution 4 ~> Idea by: Mr Morphe via https://stackoverflow.com/a/24420462
class ThousandsSeparatorRecursive < ThousandsSeparator
  def format(number, separator = ',')
    value = check_number(number)
    if value != true
      return value
    else
      format_number(number, separator)
    end
  end

  private def format_number(number, separator, number_array = [])
    # Replace negative number to string "negative"
    if number.to_i.negative?
      return '-' + format_number(number.to_s.delete('-').to_i, separator)
    end

    # Use of calc with modules + recursive if the result is positive = create loop
    if (result = number % 1000).positive?
      format_number(number / 1000, separator, number_array)
    else
      return
    end

    # concet arrays
    number_array << result

    # return (convert array to string with concat)
    number_array.join(separator)
  end
end

File: node/thousands_separator.js
/* jshint esversion: 6 */
class ThousandsSeparator {
  // Method to check the value conditions
  check_number(number) {
    // only number negative/positive is accepts or !== Sting/Integer
    if ((typeof number !== 'string' && typeof number !== 'number') || /^(|-?\d+)$/.test(number) === false) {
      return '0';
    }

    // number length < 3 does not need treatment
    if ((Math.abs(parseInt(number)).toString()).length <= 3) {
      return number.toString();
    }

    // return raise if number > 15
    if ((Math.abs(parseInt(number)).toString()).length > 15) {
      throw new Error('Number length is > 15');
    }

    // return true
    return true;
  }
}

// Solution 1
class ThousandsSeparatorRegExp extends ThousandsSeparator {
  format(number, separator = ',') {
    var value = this.check_number(number);
    if (value !== true)
      return value
    else
      return this._format_number(number, separator);
  }

  // Solution with replace via regex
  _format_number(number, separator) {
    return number.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1'+ separator).toString();
  }
}

// Solution 2
class ThousandsSeparatorArray extends ThousandsSeparator {
  format(number, separator = ',') {
    var value = this.check_number(number);
    if (value !== true)
      return value
    else
      return this._format_number(number, separator);
  }

  _format_number(number, separator) {
    var negative = false;
    var arr = [];
    var number = number.toString().split('').reverse();
  
    // Alternative with array structure
    for (var i = 0; i < number.length; i += 3) {
      arr.push(number.slice(i, i + 3))
    }
    
    // map and join in array
    var number = arr.map(function(num) {
      // for symbol 'negative'
      if (num == '-') {
        negative = true;
      }
        
      return num.reverse().join('');
    });
    
    // if exists symbol '-' then remove
    if (negative === true) {
      // remove element with symbol 'negative'
      number.splice([number.indexOf('-')], 1);
      negative = '-'
    } else {
      negative = '';
    }
    
    // return string formated
    return negative + number.reverse().join(separator).toString();
  }
}

// Solution 3 ~> Inspired by /ruby/thousands_separator.rb:68
class ThousandsSeparatorLoopString extends ThousandsSeparator {
  format(number, separator = ',') {
    var value = this.check_number(number);
    if (value !== true)
      return value
    else
      return this._format_number(number, separator);
  }

  _format_number(number, separator) {
    var value = number.toString();
    var value_size = value.length;

    // Alternative loop with slep
    for (var i = 3; i < value_size; i += 3) {
      value = value.slice(0, value_size - i) + separator + value.slice(value_size - i)
    }

    // fix for negative value
    if (value.includes('-,')) {
      value = '-' + value.slice(2, value_size + 3);
    }

    return value;
  }
}

// Solution 4 ~> Inspired by /ruby/thousands_separator.rb:105
class ThousandsSeparatorRecursive extends ThousandsSeparator {
  format(number, separator = ',') {
    var value = this.check_number(number);
    if (value !== true)
      return value
    else
      return this._format_number(number, separator);
  }

  _format_number(number, separator, number_array = []) {
    // Replace negative number to string "negative"
    if (Math.sign(parseInt(number)) === -1) {
      return '-' + this._format_number(Math.abs(number.toString()), separator)
    }

    // Use of calc with module + recursive if result is > 0 then run recursive
    let result = parseInt(number % 1000);
    if ( result > 0) {
      this._format_number(parseInt(number / 1000), separator, number_array)
    } else {
      return
    }

    // add in arrays
    number_array.push(result);

    // return (convert array to string with concat)
    return number_array.join(separator)
  }
}

module.exports = {
  ThousandsSeparatorRegExp,
  ThousandsSeparatorArray,
  ThousandsSeparatorLoopString,
  ThousandsSeparatorRecursive
}
